### PR TITLE
Improve package test coverage (build/types/parser)

### PR DESCRIPTION
**Pull request by gvallee**
**Description of the Pull Request (PR):**

Improve testing coverage for ```pkg/build/types/parser```. See issue #2559 for more details.
The coverage is not yet at 100% since i skipped coverage for the code that dealing with errors while accessing files.

**This fixes or addresses the following GitHub issues:**

- Addresses #2559 

**Before submitting a PR, make sure you have done the following:**

- Read the [Guidelines for Contributing](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md), and this PR conforms to the stated requirements.
- Added changes to the [CHANGELOG](https://github.com/sylabs/singularity/blob/master/CHANGELOG.md) if necessary according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added tests to validate this PR and tested this PR locally with a `make testall`
- Based this PR against the appropriate branch according to the [Contribution Guidelines](https://github.com/sylabs/singularity/blob/master/CONTRIBUTING.md)
- Added myself as a contributor to the [Contributors File](https://github.com/sylabs/singularity/blob/master/CONTRIBUTORS.md)

Attn: @singularity-maintainers

### CODE REVIEW

File: pkg/build/types/parser/deffile_test.go [hunk 1]
```diff
@@ -148,3 +235,144 @@ func TestParseDefinitionFileFailure(t *testing.T) {
...
+				t.Fatal("Validation of a definition file failed while expected to succeed")
+			}
+		}))
+	}
+
+	//
+	// Test with a non-existing file
+	//
+	valid, err := IsValidDefinition("notExistingDirectory/notExistingFile")
+	if valid == true && err != nil {
```
Comment (cclerget) [inline pkg/build/types/parser/deffile_test.go]: `err == nil`

File: pkg/build/types/parser/deffile_test.go [hunk 2]
```diff
@@ -148,3 +235,144 @@ func TestParseDefinitionFileFailure(t *testing.T) {
...
+		t.Fatal("Validation of a non-existing file succeeded while expected to fail")
+	}
+
+	//
+	// Test passing a valid directory in instead of a file
+	//
+	valid, err = IsValidDefinition("testdata_bad")
```
Comment (cclerget) [inline pkg/build/types/parser/deffile_test.go:352]: Looks like there is a file descriptor leak in `IsValidDefinition` otherwise `defer defFile.Close()` is not reached when definition is a directory

File: pkg/build/types/parser/deffile_test.go [hunk 3]
```diff
@@ -71,6 +73,91 @@ func TestScanDefinitionFile(t *testing.T) {
...
+	}
+	if testMap["appenv apptest"] != "test" {
+		t.Fatal("returned map is invalid", testMap["appenv"])
+	}
+}
+
+// Specific tests to cover some corner cases of doSections()
+func TestDoSections(t *testing.T) {
+	// This is an string representing an invalid section, we make sure it is not identified as a header
+	invalid_str := "%apptest\ntesttext"
```
Comment (cclerget) [inline pkg/build/types/parser/deffile_test.go]: lint complains waiting something like `invalidStr` instead

File: pkg/build/types/parser/deffile_test.go [hunk 4]
```diff
@@ -71,6 +73,91 @@ func TestScanDefinitionFile(t *testing.T) {
...
+		// Nothing to do
+	}
+
+	myerr := doSections(s1, myData)
+	if myerr == nil {
+		t.Fatal("Test passed while expected to fail")
+	}
+
+	// Now we define a valid first section but an invalid second section
+	invalid_str = "%appenv apptest apptest2\ntest\n%appenv\ntest"
```
Comment (cclerget) [inline pkg/build/types/parser/deffile_test.go]: `invalidStr`

File: pkg/build/types/parser/deffile_test.go [hunk 5]
```diff
@@ -148,3 +235,144 @@ func TestParseDefinitionFileFailure(t *testing.T) {
 		}))
 	}
 }
+
+// Specific tests to cover some corner cases of IsInvalidSectionError()
+func TestIsInvalidSectionErrors(t *testing.T) {
+
+	// Test of IsInvalidSectionError()
+	dummy_keys := []string{"dummy_key1", "dummy_key2"}
```
Comment (cclerget) [inline pkg/build/types/parser/deffile_test.go]: `dummyKeys`

File: pkg/build/types/parser/deffile_test.go [hunk 6]
```diff
@@ -148,3 +235,144 @@ func TestParseDefinitionFileFailure(t *testing.T) {
...
+
+	myerr := populateDefinition(testMap, myData)
+	if myerr != nil {
+		t.Fatal("Test failed while testing populateDefinition()")
+	}
+}
+
+// Specific tests to cover some corners cases of doHeader()
+func TestDoHeader(t *testing.T) {
+	invalidHeader := "headerTest"
```
Comment (cclerget) [inline pkg/build/types/parser/deffile_test.go]: maybe loop over `invalidHeader := []string{"headerTest", "headerTest: invalid"}` to cover both error

**Review by cclerget — CHANGES_REQUESTED** [review]

**Review by cclerget — APPROVED** [review]
